### PR TITLE
Provide capability to generate git -c calls in git commands: master

### LINKED ISSUE #5179: Git config log.showSignature breaks revision grid
**Do you want to request a *feature* or report a *bug*?**
BUG
**What is the current behavior?**
```git config log.showSignature true``` breaks the listing of all branches.  I have a few branches in my repo and when I set it, only master showed.  This is because develop had signed commits.  
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
- ```git config log.showSignature true```
- Load repo in GE

**What is the expected behavior?**
Signed commits should still show in GE.
**Environment you encounter the issue:**

- GitExtensions version:2.51.04
- GIT version: git version 2.18.0.windows.1
- OS version: Microsoft Windows [Version 6.3.9600]
- .NET version: mscorlib	DomainNeutral, Native	C:\Windows\assembly\NativeImages_v4.0.30319_64\mscorlib\be2b233163fe3b9af2d68b10d4798ef2\mscorlib.ni.dll

**Did this work in previous version of GitExtensions (which)?**
Probably not

### PR DESCRIPTION
**Pull request by vbjay**
Fixes #5179

Changes proposed in this pull request:
- Add --no-show-signature option to log calls
 
Screenshots before and after (if PR changes UI):

What did I do to test the code and ensure quality:
- Signed a commit and made sure it showed in revision grid.  
- Verified log calls were not broken

Has been tested on (remove any that don't apply):
- GIT 2.18 and above

### CODE REVIEW

**Comment by vbjay** [comment]
Willing to squash once approved.

File: GitUI/CommandsDialogs/FormVerify.LostObject.cs [hunk 1]
```diff
@@ -25,7 +25,7 @@ private sealed class LostObject
             /// %s  - subject.
             /// %ct - committer date, UNIX timestamp (easy to parse format).
             /// </summary>
-            private const string LogCommandArgumentsFormat = "log -n1 --pretty=format:\"%aN, %e, %s, %ct, %P\" {0}";
+            private const string LogCommandArgumentsFormat = "log --no-show-signature-n1 --pretty=format:\"%aN, %e, %s, %ct, %P\" {0}";
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormVerify.LostObject.cs]: ❗️ Missing a space

File: GitCommands/Git/GitGpgController.cs [hunk 2]
```diff
@@ -103,7 +103,7 @@ public GitGpgController(Func<IGitModule> getModule)
             {
                 CommitStatus cmtStatus;
 
-                string gpg = module.RunGitCmd($"log --pretty=\"format:%G?\" -1 {revision.Guid}");
+                string gpg = module.RunGitCmd($"log --no-show-signature --pretty=\"format:%G?\" -1 {revision.Guid}");
```
Comment (sharwell) [inline GitCommands/Git/GitGpgController.cs]: ❓ How is `--no-show-signature` not in conflict with `%G?`?

File: GitCommands/CommitDataManager.cs [hunk 3]
```diff
@@ -194,7 +194,7 @@ private bool TryGetCommitLog([NotNull] string commitId, [NotNull] string format,
                 return false;
             }
 
-            var arguments = $"log -1 --pretty=\"format:{format}\" {commitId}";
+            var arguments = $"log --no-show-signature -1 --pretty=\"format:{format}\" {commitId}";
```
Comment (sharwell) [inline GitCommands/CommitDataManager.cs]: ❓ What's the first version of git that accepted this argument?

Comment (sharwell) [reply GitCommands/CommitDataManager.cs]: ➡️ Git added this for 2.10.0, as mentioned in the issue for this pull request.

**Comment by sharwell** [comment]
💭 This is not the first time something like this has come up. It would be nice to have a way to ignore customized settings that affect Git output or otherwise reset things to a known default state without having to manually apply fixes like this throughout the code base.

File: GitCommands/Git/GitGpgController.cs (hunk 2 above)
Comment (vbjay) [reply GitCommands/Git/GitGpgController.cs]: Run ```git log --show-signature --pretty=%G?``` on this repo. 

Now ```git log --show-signature --pretty=%G? --no-show-signature```. Get it now?

**Comment by vbjay** [comment]
The reason this happened is git adds arguments based in config values. So our call to log was really 

git log --show-signature...  Which adds lines to the output. Have to add --no-show-signature to force it not to include the signature lines.

File: GitCommands/Git/GitGpgController.cs (hunk 2 above)
Comment (vbjay) [reply GitCommands/Git/GitGpgController.cs]: ```git log --pretty='%h %G?' | grep "G"```. Look at all the signed commits in this repo. 

Do a git log on any one of those commits. Look at what adding --show-signature does to it. Our parsing code is not expecting the gpg lines. When you have log.showsignature configured ALL log calls act like show signature is being used.

**Comment by vbjay** [comment]
Yes. That is a better solution. If you are complaining about breadth of change then you can't suggest that fix in the same breath.  I made the smallest change to solve the problem. A new issue can be raised for low level commands.

**Comment by vbjay** [comment]
The issue is that we aren't using a parser in all uses of log. You really
fail in getting it. We expect no gpg lines in the output. Putting that
argument in prevents it in case the setting is applied. We can't and
shouldn't cuss at the user. If we were using a log parser EVERYWHERE then
I'd agree. However, we are not.

**Comment by vbjay** [comment]
Not to mention avoiding the checking of is it a FPG line and avocado ding
which would slow down vs not even having the lines to begin with.

Now. I'm all for turning git log into its own class that we can turn on and
off all options at will. We would need to really do it to ALL commands

Class GitCmdParser{}

Class GitLogParser:GitCmdParser{}

Now. We obviously did not expect --show-signature output in our calls to
log. So let's say we don't want it. Why I did it the way I did. The fact
that it can sneak in with a config value means we need to guard against it.
I have done so. Later on in a major refactor we can have everything using
git log can use a git log runner/parser.

YES. I COMPLETELY AGREE that we need to use lower level commands but that
fails to work for this issue because we are not using a common framework of
reading cmd output everywhere.
....

**Comment by vbjay** [comment]
I guarantee that if I had made a common framework and altered all
runGitCmd, calls you would have complained about scope too. If you didn't
notice
https://github.com/vbjay/gitextensions/blob/5a003dbb546030e65dd096ee984243699f9bb214/GitCommands/Git/GitGpgController.cs#L106.
We have calls like this ALL over the place. So I did as little as I could
to not cause complaints on breadth of changes.

**Comment by sharwell** [comment]
Another option is to abstract the first portion of the `git` command in a way that allows us to normalize settings which are known to impact commands we are using. For example, the following would work here:

```text
git -c log.showSignature=false log ...
```

We are assuming that the base command is always just `git`, but we could define a static readonly property `BaseGitCommand` or something that returns `git -c log.showSignature=false` today and could be modified in the future to handle additional cases.

**Comment by vbjay** [comment]
Ok. Now that would fit better. I hated the parser idea because we currently
are not using a parser everywhere. Altering runGitCmd with a default that
can be managed in one spot is doable. I can implement this and it handles
future problems like this.

Could even add an overload letting user pass in a config options object
that could then be turned into -c calls.

RunGitCmd("log ...", new CfgOpts( {ShowNoSignature=true}))

Something like above.

**Comment by sharwell** [comment]
:memo: The previous could be done as part of the central command execution without updating all the call sites if we recognize command of the form "git [command]" and rewrite them to "git [overrides] [command]" before starting the process.

**Comment by vbjay** [comment]
Right. But I am thinking of the defaults not being right for a specific
call. Yes. The default would just be like the calls are now. Still allowing
a caller to pass in new defaults that override the default values or even
add a new config value would help.

**Comment by sharwell** [comment]
❓ Why would we not want the defaults to be like the calls are today *for me* (where it works)? In other words, what harm is there in normalizing potentially problematic options for all internal calls? After all, it wouldn't affect commands run explicitly from the Console tab or commands run from any other environment.

**Comment by sharwell** [comment]
💭 Would need to ensure it didn't cause problems for `git config` calls we make (if any). Maybe it's not a problem but would want to be sure.

**Comment by vbjay** [comment]
You're not getting it. The defaults would handle 99% of the calls.  I'm
thinking of the caller that would want showsignature or other option.   The
runGitCmd would start with defaults but use could add options to it that
override the defaults.

Frankly the default would end up calling this overload.

RunGitCmd("log ...") Would call RunGitCmd("log ...', theDefaults) which
would actually run

git -c... -c... log ...

**Comment by vbjay** [comment]
Actually the defaults could be an
Ienumerable<keyvaluepair<string,string>>.

It would be simple to generate -c calls out of them. This way if the
default and the user calling with an override set the same config, the
override would win. I have it figured out. I'll get it done and you will
get it.

**Comment by vbjay** [comment]
-c is just one more layer of git config. Just like --global is overriden by
git config. It wouldn't cause a problem. We just want to be lean with our
defaults. Don't want to override too much.

**Comment by vbjay** [comment]
Verified that ```git -c log.ShowSignature=false log --pretty='%h %G?'```
does fix the problem.

File: GitCommands/CommitDataManager.cs (hunk 3 above)
Comment (RussKie) [reply GitCommands/CommitDataManager.cs]: @gerhardol we'll need to bump the min version to 2.10

Comment (RussKie) [reply GitCommands/CommitDataManager.cs]: I'd like to avoid patching 2.51 unless absolutely necessary

**Comment by drewnoakes** [comment]
I haven't looked deeply at this (though have wondered before whether `-c` options might make good defaults) but one idea work a look is to set the `GIT_CONFIG` environment variable on the `git.exe` process and see if we can null out external user configuration somehow. In that way we can set a bunch of standard configuration for all commands in one place.

**Comment by vbjay** [comment]
I'd rather stick with -c.  I am putting it in the RunGitCmd and related
methods.  This way the caller can throw in their config needs if needed.
Example is the customized -c calls for diff.  We can default specific
values like the showsignature and if we need extra arguments the caller can
call the override passing in their overrides.  If the caller passes in a
value that is managed by teh defaults, it will override the default.

**Comment by vbjay** [comment]
The other reason why I want to use -c is the log shows exactly what is being run.

**Comment by drewnoakes** [comment]
> The other reason why I want to use -c is the log shows exactly what is being run.

Good point. It's nice to be able to copy/paste the command locally and have the same output.

File: GitCommands/Git/GitModule.cs [hunk 4]
```diff
@@ -2464,7 +2528,12 @@ public string GetStatusText(bool untracked)
         public string GetDiffFilesText(string firstRevision, string secondRevision, bool noCache = false)
         {
             string cmd = DiffCommandWithStandardArgs + "-M -C --name-status " + _revisionDiffProvider.Get(firstRevision, secondRevision);
-            return noCache ? RunGitCmd(cmd) : RunCacheableGitCmd(cmd, SystemEncoding);
+            KeyValuePair<string, string>[] configs = new KeyValuePair<string, string>[]
```
Comment (gerhardol) [inline GitCommands/Git/GitModule.cs]: Can this be broken out so this is still centralized.
Now the config is repeated four times.

**Comment by vbjay** [comment]
Yes. I was just showing the use of it. I can create a diffConfig variable.

**Comment by vbjay** [comment]
Also I'll prevent nulls and lowercase the keys.

**Comment by vbjay** [comment]
I'll squash when approved

**Comment by vbjay** [comment]
Rebased on master.

**Comment by vbjay** [comment]
Rebased on master.

**Comment by drewnoakes** [comment]
On my phone now but will give this a review soon as possible.

**Comment by vbjay** [comment]
Please just approve if good. I'll squash as the last step.

**Comment by drewnoakes** [comment]
I have a few thoughts here, but the initial point is that this applies to every git command which seems like overkill. It's harmless, but clutter. Presumably the idea is to add more defaults over time, increasing this.

image

I would like to see the number of methods on `GitModule` that launch processes reduced rather than increased. There's a [branch](https://github.com/drewnoakes/gitextensions/tree/git-executable) I'm working on that should get us down to a single method for this. As discussed in #4643 we could then locate all invocations of `git log` into a single place and apply the configuration there.

My feeling is that we can find a simple approach that achieves the following goals:

1. Fix the bug
1. Don't require changes sprawled throughout the code
1. Keep the API simple and focussed

In its simplest form, something like this would suffice, piggy backing off `ArgumentBuilder`:

```c#
public string Log(ObjectId commitId)
{
    return RunGitCmd(new ArgumentBuilder
    {
        "-c log.showSignature=false",
        "log",
        "-n 1",
        commitId
```

@vbjay I'm interested to hear how else you would anticipate these new APIs being used.

Apologies for the quick review but I have to run. I'll be back online later in the day to respond to any thoughts.

**Comment by vbjay** [comment]
1:  It does.  

2: It is not sprawled anywhere.  If the caller doesn't want to add any configs it is as it always was.  They don't call the overload.  I will not support just plain strings.  The whole point was to have OVERRIDABLE configs.  The default that all commands get and then (additional + overriding configs).  The user shouldn't have to remember where the -c calls go.  All the calls except for the diff calls I modified use teh defaults and were not modified.  I will not support just plain -c strings in the argument builder.  Too easy to put it in the wrong spot.  



3:  It is simple.  I did it this way to force the users to define the config keys and even describe why the config key is going in.  Yes.  The config keypair array code is verbose.  IT SHOULD be.  
```c#
private static KeyValuePair<string, string>[] diffConfigs = new KeyValuePair<string, string>[]
             {
             new KeyValuePair<string, string>("diff.submodule", "short"),// Why this config
             new KeyValuePair<string, string>("diff.noprefix", "false")// Why this config
             };
```


Let's say the programmer is coding a plugin and wants th log.showsignature  All they have to do is call the overload passing in 
```c#
new KeyValuePair<string, string>[]
             {
             new KeyValuePair<string, string>("log.ShowSignature", "true")//Need gpg info in commit log
             };
```

The -c calls are not command arguments.  They are git config changes and not part of the arguments of the specific git command.  They should be separate.  They are.

**Comment by drewnoakes** [comment]
I didn't say the proposed changes don't meet goals 1 and 2, just that the simple alternative I show also achieves the same. Overriding might be useful. I'll think about it some more. Do you have any idea how to avoid adding defaults to all operations?

>  log.showsignature does not just affect git log

The docs list three commands it does apply to. Can the code filter to just those commands?

**Comment by vbjay** [comment]
> The docs list three commands it does apply to. Can the code filter to just those commands?

The only way to really do that reliably and well is to turn every git command into a class.  Do you want to suss out log or whatever from the git command?

```c#

Class GitCommand
{
...
}

Class LogCommand: GitCommand
{
...
}
```

**Comment by vbjay** [comment]
One thing that could be done is to hide the -c part in the log viewer and only include it when being copied.  Showing the configs in a listbox or other ui on the logviewer when the command is selected.

**Comment by drewnoakes** [comment]
> Do you want to suss out log or whatever from the git command?

We don't have to do any parsing as all the input (AFAIK) comes from our code itself. We can just change the calling code. Maybe scripts are an exception.

> The only way to really do that reliably and well is to turn every git command into a class

That is one way of doing it. We can use data instead of types. Type hierarchies tend to contribute to the sprawl we mentioned earlier.

```c#
public string Log(ObjectId commitId)
{
    return RunGitCmd(new GitArgumentBuilder("log")
    {
        new GitConfigItem("log.showSignature", "overriding value"),
        "-n 1",
        commitId
    });
}
```

`GitArgumentBuilder` can have a map from command string to zero or more config items. These would then be overridden by subsequent `Add` calls. I think it will be handy to be able to pass such a builder around to build up argument lists, and by differentiating the config with its own type, we can sort out overrides and position the `-c` items before the command when `ToString` is called, right before starting the process.

```c#
public sealed class GitArgumentBuilder
{
    private static readonly Dictionary<string, IReadOnlyList<(string key, string value)>>
        _configItemsByCommand = new Dictionary<string, IReadOnlyList<(string key, string value)>>
        {
            {
                "log",
                new[]
                {
                    ("log.showSignature", "false"),
                    ("log.abbrevCommit", "false")
                },
            }
            { "show", new[] { ("log.showSignature", "false") } },
            { "whatchanged", new[] { ("log.showSignature", "false") } },
            { "rebase", new[] { ("rebase.autoSquash", "false") } },
            // ...
        };

    private readonly List<(string key, string value)> _configItems = new List<(string key, string value)>();
    private readonly StringBuilder _arguments = new StringBuilder();
    private readonly string _command;

    public GitArgumentBuilder(string command)
    {
        _command = command;
    }

    public void Add(string argument) { ... }
    public void Add(GitConfigItem configItem) { ... }
    public override string ToString() { ... }
}
```

Another complementary idea is to make `RunGitCmd` accept some type that has implicit conversions from `string`, `ArgumentBuilder`, `GitArgumentBuilder` and whatever else we end up with. We can avoid overloads this way and keep the construction of executable arguments separate from the management of processes. `ArgumentBuilder` is currently agnostic to any executable's expectations around argument lists. The work on my branch that I mentioned earlier is also exploring the fact that we launch and manage other executable processes, not just `git.exe`.

**Comment by RussKie** [comment]
> The only way to really do that reliably and well is to turn every git command into a class. 

I actually tend to favour this idea - this way we can have small (tested) implementations with single responsibilities.

**Comment by drewnoakes** [comment]
> I actually tend to favour this idea

Could you sketch out a few lines of code to show what sort of thing you mean?

In #4643 there was an idea to split `GitModule` up into smaller classes based around different kinds of operations you might perform against a module, such as rebasing, committing, logging, etc. I have no problem with that. I don't think it's worth creating an object hierarchy just for the formulation of git argument lists, however. This can adequately be done inside types such as `ISubmoduleCommands` implementations, with small clear tests.

Some loose pseudo code of what I'm shooting for:

```c#
// In SubmoduleCommandsTests
[Test]
public void TestSubmoduleUpdate()
{
    var executable = new MockGitExecutable();
    
    var commands = new SubmoduleCommands(executable, module);

    // stage
    executable.StageGitCommand(
        "submodule update --some-argument abc", 
        "git process output here");

    // act
    commands.Update("abc", foo, bar);

    // verify
    executable.Verify();
}
```

The process stuff is coming along nicely. If we land that then I'd like to get back to the `GitModule` refactor, which opens up easier testing of all git operations as well as consolidating related code. We might even be able to prohibit arbitrary forms/code from firing off arbitrary git operations, instead having to go through one of the command family objects.

**Comment by vbjay** [comment]
Yes. Well I don't thnk that falls in the scope of fixing this issue. If be
more interested in seeing that after Drew's work is merged.

I do agree.

**Comment by vbjay** [comment]
I do like the idea of cmdx having config values.  

```c#
public sealed class GitArgumentBuilder
{
    private static readonly Dictionary<string, IReadOnlyList<(string key, string value)>>
        _configItemsByCommand = new Dictionary<string, IReadOnlyList<(string key, string value)>>
        {
            {
                "log",
                new[]
                {
                    ("log.showSignature", "false"),
                    ("log.abbrevCommit", "false")
                },
            }
            { "show", new[] { ("log.showSignature", "false") } },
            { "whatchanged", new[] { ("log.showSignature", "false") } },
            { "rebase", new[] { ("rebase.autoSquash", "false") } },
            // ...
        };
...
```
That idea does have merit.  I think separating command(log, diff...) from the arguments like you suggest here would help.  Of course that would mean a huge change of a TON of files and someone complained of breadth of changes a while ago.

**Comment by vbjay** [comment]
Good point on the rebase.autosquash config value when we show a checkbox that controls that argument.

**Comment by RussKie** [comment]
> Could you sketch out a few lines of code 

The idea I favour is a command pattern, each command we execute is encapsulated in own class. This way we can test each command.
Of course it means we probably end up with more code, but I think in this case it may be justified.

**Comment by vbjay** [comment]
@RussKie  do you agree that a change like that should be its own issue and pull request?  A rework like that  would require a lot. I'm trying to solve the -c part of calling a git command with overrides as a possibility. I do like Drew's idea but I think that should happen after we rework to commands being classes.

**Comment by RussKie** [comment]
Absolutely agree, I didn't say I expected this to be as part of this PR.

**Comment by drewnoakes** [comment]
@vbjay please take a look at the ideas in #5281, specifically `IExecutable` and `ArgumentString`. I believe a `GitCommandBuilder` class could derive from `ArgumentBuilder` and provide the logic you are aiming for here in a way that composes really nicely across all operations. It would override the `ToString` method to apply config arguments ahead of the command and regular argument list.

That is orthogonal to a command pattern for git operations. Such and idea could still be implemented in a clean fashion on top of that PR, with or without a `GitCommandBuilder`.

I would still like to see some rough API ideas for such a pattern just to see how such an abstraction would work. We already have `IGitCommand`, but it only has four implementations. I agree that such a change is not within the scope of what this PR is setting out to achieve. Perhaps once an idea is more fleshed out a discussion issue could be created.

**Comment by vbjay** [comment]
I will. Sorry. I didn't see this earlier.

**Comment by drewnoakes** [comment]
@vbjay, no worries. I put together some code that does what was discussed above, based on top of #5281 in https://github.com/drewnoakes/gitextensions/pull/6.

**Comment by vbjay** [comment]
Also. We could just have a static class for default configs and have a
lookup. Rough idea (on mobile excuse rough code)

Class DefaultGitConfigs
{
Config[] configs = new Config[] {new
Config("log","log.showsignature","false")};

Public static IEnumerable GetCommandConfigs(string cmd)
=> configs.ToLookup(c=> c.command)(cmd)

...

**Comment by drewnoakes** [comment]
> I still want to use something like a keyvaluepair or a class that holds the values for config.

Agree. drewnoakes#6 introduces `GitConfigItem` which is a readonly struct over `Key`/`Value` sting properties. It's structurally equivalent to `KeyValuePair<string,string>` but more semantically focussed.

This file shows the moving parts:

https://github.com/drewnoakes/gitextensions/blob/git-argument-builder/GitExtUtils/GitArgumentBuilder.cs

Those changes take care of overrides within the builder.

> We could just have a static class for default configs and have a lookup

Take a look at `GitCommandConfiguration` in the file linked above. It has a static `Default` property that serves an instance of itself configured for GE purposes. However it's an instance type so we can test different scenarios with different config as needed by passing an instance into `GitArgumentBuilder`. It could be specified per module, for example. It has an API similar to what you propose: `IReadOnlyList<GitConfigItem> Get(string command)`

**Pull request closed without merging**